File: linux/src/sensors.py
"""
Hardware sensor collectors for TaskbarMonitor (Linux).

Everything is read from /proc and /sys (world-readable) or from nvidia-smi.
No root, no kernel module: on Linux the temperatures the Windows build needed a
driver for are already exposed under /sys/class/hwmon and /sys/block/*/device/hwmon.

Sensors.read() returns an immutable Snapshot. All fields are Optional — a missing
sensor is None, never an exception.
"""
from __future__ import annotations

import glob
import os
import shutil
import subprocess
import time
from dataclasses import dataclass, field


@dataclass(frozen=True)
class Disk:
    name: str
    load: float | None          # % busy
    temp: float | None          # °C


@dataclass(frozen=True)
class Snapshot:
    cpu_load: float | None = None
    cpu_temp: float | None = None
    gpu_load: float | None = None
    gpu_temp: float | None = None
    gpu_name: str | None = None
    ram_load: float | None = None
    ram_used_gb: float | None = None
    ram_total_gb: float | None = None
    disks: tuple[Disk, ...] = ()
    net_up_bps: float | None = None
    net_down_bps: float | None = None


def _read_int(path: str) -> int | None:
    try:
        with open(path) as f:
            return int(f.read().strip())
    except (OSError, ValueError):
        return None


def _read_str(path: str) -> str | None:
    try:
        with open(path) as f:
            return f.read().strip()
    except OSError:
        return None


class Sensors:
    """Stateful collector: holds previous counters to compute rates/deltas."""

    def __init__(self) -> None:
        self._t_prev: float | None = None
        self._cpu_prev: tuple[int, int] | None = None      # (idle, total)
        self._net_prev: tuple[int, int] | None = None      # (rx, tx)
        self._disk_prev: dict[str, int] = {}               # name -> io_ms
        self._cpu_hwmon: str | None = None                 # cached path
        self._cpu_hwmon_scanned = False
        self._gpu_kind: str | None = None                  # 'nvidia'|'amd'|'intel'|'none'

    # ---- public -------------------------------------------------------------

    def read(self) -> Snapshot:
        now = time.monotonic()
        dt = (now - self._t_prev) if self._t_prev else None
        self._t_prev = now

        cpu_load = self._cpu_load()
        cpu_temp = self._cpu_temp()
        ram_load, ram_used, ram_total = self._memory()
        disks = self._disks(dt)
        up, down = self._network(dt)
        gpu_load, gpu_temp, gpu_name = self._gpu()

        return Snapshot(
            cpu_load=cpu_load, cpu_temp=cpu_temp,
            gpu_load=gpu_load, gpu_temp=gpu_temp, gpu_name=gpu_name,
            ram_load=ram_load, ram_used_gb=ram_used, ram_total_gb=ram_total,
            disks=disks, net_up_bps=up, net_down_bps=down,
        )

    # ---- CPU ----------------------------------------------------------------

    def _cpu_load(self) -> float | None:
        try:
            with open("/proc/stat") as f:
                parts = f.readline().split()
            if parts[0] != "cpu":
                return None
            vals = [int(x) for x in parts[1:]]
        except (OSError, ValueError, IndexError):
            return None
        idle = vals[3] + (vals[4] if len(vals) > 4 else 0)   # idle + iowait
        total = sum(vals)
        prev = self._cpu_prev
        self._cpu_prev = (idle, total)
        if prev is None:
            return None
        d_idle = idle - prev[0]
        d_total = total - prev[1]
        if d_total <= 0:
            return None
        return max(0.0, min(100.0, (1.0 - d_idle / d_total) * 100.0))

    def _cpu_temp(self) -> float | None:
        if not self._cpu_hwmon_scanned:
            self._cpu_hwmon = self._find_cpu_hwmon()
            self._cpu_hwmon_scanned = True
        if not self._cpu_hwmon:
            return None
        return self._best_cpu_temp(self._cpu_hwmon)

    @staticmethod
    def _find_cpu_hwmon() -> str | None:
        # Prefer, in order: coretemp (Intel), k10temp/zenpower (AMD),
        # then generic thermal zones.
        order = ["coretemp", "k10temp", "zenpower", "cpu_thermal", "acpitz"]
        found: dict[str, str] = {}
        for h in glob.glob("/sys/class/hwmon/hwmon*"):
            name = _read_str(os.path.join(h, "name"))
            if name and name in order and name not in found:
                found[name] = h
        for name in order:
            if name in found:
                return found[name]
        return None

    @staticmethod
    def _best_cpu_temp(hwmon: str) -> float | None:
        # Prefer the package/Tctl label; else the max core reading.
        best_rank, best_val = 99, None
        for inp in glob.glob(os.path.join(hwmon, "temp*_input")):
            raw = _read_int(inp)
            if raw is None:
                continue
            val = raw / 1000.0
            if not (0 < val < 150):
                continue
            label = _read_str(inp.replace("_input", "_label")) or ""
            rank = 3
            if label.startswith("Package id") or label in ("Tctl", "Tdie"):
                rank = 0
            elif label.startswith("Core") or label.startswith("Tccd"):
                rank = 1
            if rank < best_rank or (rank == best_rank and (best_val is None or val > best_val)):
                best_rank, best_val = rank, val
        return best_val

    # ---- memory -------------------------------------------------------------

    @staticmethod
    def _memory() -> tuple[float | None, float | None, float | None]:
        info: dict[str, int] = {}
        try:
            with open("/proc/meminfo") as f:
                for line in f:
                    k, _, rest = line.partition(":")
                    info[k] = int(rest.strip().split()[0])  # kB
        except (OSError, ValueError, IndexError):
            return None, None, None
        total = info.get("MemTotal")
        avail = info.get("MemAvailable")
        if not total:
            return None, None, None
        if avail is None:
            avail = info.get("MemFree", 0)
        used = total - avail
        total_gb = total / 1048576.0
        used_gb = used / 1048576.0
        return used / total * 100.0, used_gb, total_gb

    # ---- disks --------------------------------------------------------------

    def _disks(self, dt: float | None) -> tuple[Disk, ...]:
        result: list[Disk] = []
        for dev in sorted(self._block_devices()):
            io_ms = self._disk_io_ms(dev)
            load = None
            if io_ms is not None:
                prev = self._disk_prev.get(dev)
                self._disk_prev[dev] = io_ms
                if prev is not None and dt and dt > 0:
                    load = max(0.0, min(100.0, (io_ms - prev) / (dt * 1000.0) * 100.0))
            result.append(Disk(name=self._disk_label(dev), load=load, temp=self._disk_temp(dev)))
        return tuple(result)

    @staticmethod
    def _block_devices() -> list[str]:
        devs = []
        for path in glob.glob("/sys/block/*"):
            dev = os.path.basename(path)
            if dev.startswith(("loop", "ram", "zram", "dm-", "md", "sr")):
                continue
            devs.append(dev)
        return devs

    @staticmethod
    def _disk_io_ms(dev: str) -> int | None:
        stat = _read_str(f"/sys/block/{dev}/stat")
        if not stat:
            return None
        f = stat.split()
        # field 10 (1-based) of block stat = time spent doing I/Os (ms)
        try:
            return int(f[9])
        except (IndexError, ValueError):
            return None

    @staticmethod
    def _disk_label(dev: str) -> str:
        model = _read_str(f"/sys/block/{dev}/device/model")
        return model if model else dev

    @staticmethod
    def _disk_temp(dev: str) -> float | None:
        # nvme and drivetemp (SATA) expose temp under the device's hwmon.
        best = None
        for inp in glob.glob(f"/sys/block/{dev}/device/hwmon/hwmon*/temp*_input"):
            raw = _read_int(inp)
            if raw is None:
                continue
            val = raw / 1000.0
            if not (0 < val < 120):
                continue
            label = _read_str(inp.replace("_input", "_label")) or ""
            # skip alarm/threshold pseudo-sensors if labelled
            if "Composite" in label or label == "":
                return val
            if best is None:
                best = val
        return best

    # ---- network ------------------------------------------------------------

    def _network(self, dt: float | None) -> tuple[float | None, float | None]:
        rx_total = tx_total = 0
        any_if = False
        for iface in self._net_interfaces():
            rx = _read_int(f"/sys/class/net/{iface}/statistics/rx_bytes")
            tx = _read_int(f"/sys/class/net/{iface}/statistics/tx_bytes")
            if rx is None or tx is None:
                continue
            rx_total += rx
            tx_total += tx
            any_if = True
        if not any_if:
            return None, None
        prev = self._net_prev
        self._net_prev = (rx_total, tx_total)
        if prev is None or not dt or dt <= 0:
            return None, None
        down = max(0.0, (rx_total - prev[0]) / dt)
        up = max(0.0, (tx_total - prev[1]) / dt)
        return up, down

    @staticmethod
    def _net_interfaces() -> list[str]:
        ifaces = []
        for path in glob.glob("/sys/class/net/*"):
            iface = os.path.basename(path)
            if iface == "lo":
                continue
            # real hardware interfaces have a /device symlink; this cleanly
            # excludes veth/bridge/docker/virbr/tap/tun virtual interfaces
            if os.path.exists(os.path.join(path, "device")):
                ifaces.append(iface)
        if ifaces:
            return ifaces
        # fallback: everything that isn't loopback or an obvious virtual iface
        skip = ("lo", "veth", "br-", "docker", "virbr", "vnet", "tap", "tun", "wg", "zt")
        return [
            os.path.basename(p) for p in glob.glob("/sys/class/net/*")
            if not os.path.basename(p).startswith(skip)
        ]

    # ---- GPU ----------------------------------------------------------------

    def _gpu(self) -> tuple[float | None, float | None, str | None]:
        if self._gpu_kind is None:
            self._gpu_kind = self._detect_gpu()
        if self._gpu_kind == "nvidia":
            return self._gpu_nvidia()
        if self._gpu_kind == "amd":
            return self._gpu_amd()
        if self._gpu_kind == "intel":
            return self._gpu_intel()
        return None, None, None

    def _detect_gpu(self) -> str:
        if shutil.which("nvidia-smi") and self._gpu_nvidia()[0] is not None:
            return "nvidia"
        if glob.glob("/sys/class/drm/card*/device/gpu_busy_percent"):
            return "amd"
        if glob.glob("/sys/class/drm/card*/device/hwmon/hwmon*/temp*_input"):
            return "intel"
        return "none"

    @staticmethod
    def _gpu_nvidia() -> tuple[float | None, float | None, str | None]:
        try:
            out = subprocess.run(
                ["nvidia-smi",
                 "--query-gpu=utilization.gpu,temperature.gpu,name",
                 "--format=csv,noheader,nounits"],
                capture_output=True, text=True, timeout=2, check=False,
            ).stdout.strip().splitlines()
            if not out:
                return None, None, None
            load_s, temp_s, name = (x.strip() for x in out[0].split(",", 2))
            return float(load_s), float(temp_s), name
        except (OSError, ValueError, subprocess.SubprocessError):
            return None, None, None

    @staticmethod
    def _gpu_amd() -> tuple[float | None, float | None, str | None]:
        load = temp = None
        cards = sorted(glob.glob("/sys/class/drm/card*/device/gpu_busy_percent"))
        if cards:
            v = _read_int(cards[0])
            load = float(v) if v is not None else None
            base = os.path.dirname(cards[0])
            for inp in glob.glob(os.path.join(base, "hwmon/hwmon*/temp*_input")):
                raw = _read_int(inp)
                if raw is not None and 0 < raw / 1000.0 < 150:
                    temp = raw / 1000.0
                    break
        return load, temp, "AMD GPU"

    @staticmethod
    def _gpu_intel() -> tuple[float | None, float | None, str | None]:
        for inp in glob.glob("/sys/class/drm/card*/device/hwmon/hwmon*/temp*_input"):
            raw = _read_int(inp)
            if raw is not None and 0 < raw / 1000.0 < 150:
                return None, raw / 1000.0, "Intel GPU"
        return None, None, "Intel GPU"

File: linux/gnome-extension/taskbar-monitor@marllongomes.github.io/sensors.js
/*
 * Hardware sensor collectors for TaskbarMonitor (GNOME Shell extension).
 *
 * Everything is read from /proc and /sys (world-readable) or from nvidia-smi.
 * No root, no kernel module. All fields of a snapshot are null when a sensor
 * is missing — never an exception.
 *
 * GJS port of linux/src/sensors.py; keep the two in sync.
 */
import GLib from 'gi://GLib';
import Gio from 'gi://Gio';

Gio._promisify(Gio.Subprocess.prototype, 'communicate_utf8_async');

const decoder = new TextDecoder('utf-8');

function readStr(path) {
    try {
        const [ok, bytes] = GLib.file_get_contents(path);
        return ok ? decoder.decode(bytes).trim() : null;
    } catch {
        return null;
    }
}

function readInt(path) {
    const s = readStr(path);
    if (s === null)
        return null;
    const v = parseInt(s, 10);
    return Number.isNaN(v) ? null : v;
}

function listDir(path) {
    const names = [];
    try {
        const en = Gio.File.new_for_path(path).enumerate_children(
            'standard::name', Gio.FileQueryInfoFlags.NONE, null);
        let info;
        while ((info = en.next_file(null)) !== null)
            names.push(info.get_name());
        en.close(null);
    } catch {
        // missing directory -> no entries
    }
    return names.sort();
}

function exists(path) {
    return GLib.file_test(path, GLib.FileTest.EXISTS);
}

/** Stateful collector: holds previous counters to compute rates/deltas. */
export class Sensors {
    constructor() {
        this._tPrev = null;
        this._cpuPrev = null;              // [idle, total]
        this._netPrev = null;              // [rx, tx]
        this._diskPrev = new Map();        // dev -> io_ms
        this._cpuHwmon = undefined;        // undefined = not scanned yet
        this._diskHwmon = new Map();       // dev -> hwmon dir or null
        this._gpuKind = null;              // 'nvidia'|'amd'|'intel'|'none'
        this._nvidiaMisses = 0;
    }

    /** Async: nvidia-smi (when present) runs as a subprocess. */
    async sample() {
        const now = GLib.get_monotonic_time() / 1e6;
        const dt = this._tPrev !== null ? now - this._tPrev : null;
        this._tPrev = now;

        const [ramLoad, ramUsedGb, ramTotalGb] = this._memory();
        const [netUp, netDown] = this._network(dt);
        const [gpuLoad, gpuTemp, gpuName] = await this._gpu();

        return {
            cpuLoad: this._cpuLoad(),
            cpuTemp: this._cpuTemp(),
            gpuLoad, gpuTemp, gpuName,
            ramLoad, ramUsedGb, ramTotalGb,
            disks: this._disks(dt),        // [{name, dev, load, temp}]
            netUpBps: netUp,
            netDownBps: netDown,
        };
    }

    // ---- CPU ---------------------------------------------------------------

    _cpuLoad() {
        const stat = readStr('/proc/stat');
        if (!stat)
            return null;
        const parts = stat.split('\n', 1)[0].split(/\s+/);
        if (parts[0] !== 'cpu')
            return null;
        const vals = parts.slice(1).map(Number);
        if (vals.some(Number.isNaN) || vals.length < 4)
            return null;
        const idle = vals[3] + (vals.length > 4 ? vals[4] : 0);   // idle + iowait
        const total = vals.reduce((a, b) => a + b, 0);
        const prev = this._cpuPrev;
        this._cpuPrev = [idle, total];
        if (!prev)
            return null;
        const dIdle = idle - prev[0];
        const dTotal = total - prev[1];
        if (dTotal <= 0)
            return null;
        return Math.max(0, Math.min(100, (1 - dIdle / dTotal) * 100));
    }

    _cpuTemp() {
        if (this._cpuHwmon === undefined)
            this._cpuHwmon = this._findCpuHwmon();
        if (!this._cpuHwmon)
            return null;
        return this._bestCpuTemp(this._cpuHwmon);
    }

    _findCpuHwmon() {
        // Prefer, in order: coretemp (Intel), k10temp/zenpower (AMD),
        // then generic thermal zones.
        const order = ['coretemp', 'k10temp', 'zenpower', 'cpu_thermal', 'acpitz'];
        const found = {};
        for (const entry of listDir('/sys/class/hwmon')) {
            const dir = `/sys/class/hwmon/${entry}`;
            const name = readStr(`${dir}/name`);
            if (name && order.includes(name) && !(name in found))
                found[name] = dir;
        }
        for (const name of order) {
            if (name in found)
                return found[name];
        }
        return null;
    }

    _bestCpuTemp(hwmon) {
        // Prefer the package/Tctl label; else the max core reading.
        let bestRank = 99, bestVal = null;
        for (const entry of listDir(hwmon)) {
            if (!/^temp\d+_input$/.test(entry))
                continue;
            const raw = readInt(`${hwmon}/${entry}`);
            if (raw === null)
                continue;
            const val = raw / 1000;
            if (!(val > 0 && val < 150))
                continue;
            const label = readStr(`${hwmon}/${entry.replace('_input', '_label')}`) ?? '';
            let rank = 3;
            if (label.startsWith('Package id') || label === 'Tctl' || label === 'Tdie')
                rank = 0;
            else if (label.startsWith('Core') || label.startsWith('Tccd'))
                rank = 1;
            if (rank < bestRank || (rank === bestRank && (bestVal === null || val > bestVal))) {
                bestRank = rank;
                bestVal = val;
            }
        }
        return bestVal;
    }

    // ---- memory ------------------------------------------------------------

    _memory() {
        const text = readStr('/proc/meminfo');
        if (!text)
            return [null, null, null];
        const info = {};
        for (const line of text.split('\n')) {
            const m = line.match(/^(\w+):\s+(\d+)/);
            if (m)
                info[m[1]] = parseInt(m[2], 10);          // kB
        }
        const total = info['MemTotal'];
        if (!total)
            return [null, null, null];
        const avail = info['MemAvailable'] ?? info['MemFree'] ?? 0;
        const used = total - avail;
        return [used / total * 100, used / 1048576, total / 1048576];
    }

    // ---- disks -------------------------------------------------------------

    _disks(dt) {
        const result = [];
        for (const dev of this._blockDevices()) {
            const ioMs = this._diskIoMs(dev);
            let load = null;
            if (ioMs !== null) {
                const prev = this._diskPrev.get(dev);
                this._diskPrev.set(dev, ioMs);
                if (prev !== undefined && dt && dt > 0)
                    load = Math.max(0, Math.min(100, (ioMs - prev) / (dt * 1000) * 100));
            }
            result.push({
                name: this._diskLabel(dev),
                dev,
                load,
                temp: this._diskTemp(dev),
            });
        }
        return result;
    }

    _blockDevices() {
        return listDir('/sys/block').filter(dev =>
            !/^(loop|ram|zram|dm-|md|sr)/.test(dev));
    }

    _diskIoMs(dev) {
        const stat = readStr(`/sys/block/${dev}/stat`);
        if (!stat)
            return null;
        const f = stat.split(/\s+/).filter(s => s.length);
        // field 10 (1-based) of block stat = time spent doing I/Os (ms)
        const v = parseInt(f[9], 10);
        return Number.isNaN(v) ? null : v;
    }

    _diskLabel(dev) {
        const model = readStr(`/sys/block/${dev}/device/model`);
        return model || dev;
    }

    _diskHwmonDir(dev) {
        if (this._diskHwmon.has(dev))
            return this._diskHwmon.get(dev);
        // nvme and drivetemp (SATA) expose temp under a hwmon near the device.
        let found = null;
        for (const base of [`/sys/block/${dev}/device/hwmon`,     // SATA drivetemp
            `/sys/block/${dev}/device`]) {                        // NVMe: hwmonN dirs
            if (!exists(base))
                continue;
            for (const entry of listDir(base)) {
                if (/^hwmon\d+$/.test(entry) && exists(`${base}/${entry}`)) {
                    found = `${base}/${entry}`;
                    break;
                }
            }
            if (found)
                break;
        }
        this._diskHwmon.set(dev, found);
        return found;
    }

    _diskTemp(dev) {
        const hwmon = this._diskHwmonDir(dev);
        if (!hwmon)
            return null;
        let best = null;
        for (const entry of listDir(hwmon)) {
            if (!/^temp\d+_input$/.test(entry))
                continue;
            const raw = readInt(`${hwmon}/${entry}`);
            if (raw === null)
                continue;
            const val = raw / 1000;
            if (!(val > 0 && val < 120))
                continue;
            const label = readStr(`${hwmon}/${entry.replace('_input', '_label')}`) ?? '';
            if (label.includes('Composite') || label === '')
                return val;
            if (best === null)
                best = val;
        }
        return best;
    }

    // ---- network -----------------------------------------------------------

    _network(dt) {
        let rxTotal = 0, txTotal = 0, anyIf = false;
        for (const iface of this._netInterfaces()) {
            const rx = readInt(`/sys/class/net/${iface}/statistics/rx_bytes`);
            const tx = readInt(`/sys/class/net/${iface}/statistics/tx_bytes`);
            if (rx === null || tx === null)
                continue;
            rxTotal += rx;
            txTotal += tx;
            anyIf = true;
        }
        if (!anyIf)
            return [null, null];
        const prev = this._netPrev;
        this._netPrev = [rxTotal, txTotal];
        if (!prev || !dt || dt <= 0)
            return [null, null];
        return [
            Math.max(0, (txTotal - prev[1]) / dt),   // up
            Math.max(0, (rxTotal - prev[0]) / dt),   // down
        ];
    }

    _netInterfaces() {
        const all = listDir('/sys/class/net').filter(i => i !== 'lo');
        // real hardware interfaces have a /device symlink; this cleanly
        // excludes veth/bridge/docker/virbr/tap/tun virtual interfaces
        const hw = all.filter(i => exists(`/sys/class/net/${i}/device`));
        if (hw.length)
            return hw;
        const skip = /^(veth|br-|docker|virbr|vnet|tap|tun|wg|zt)/;
        return all.filter(i => !skip.test(i));
    }

    // ---- GPU ---------------------------------------------------------------

    async _gpu() {
        if (this._gpuKind === null)
            this._gpuKind = this._detectGpu();
        if (this._gpuKind === 'nvidia') {
            const res = await this._gpuNvidia();
            // nvidia-smi present but persistently yields nothing (e.g. dGPU
            // powered off with no driver): fall back to the sysfs paths.
            if (res[0] === null && res[1] === null) {
                if (++this._nvidiaMisses >= 3)
                    this._gpuKind = this._detectGpuSysfs();
            } else {
                this._nvidiaMisses = 0;
            }
            return res;
        }
        if (this._gpuKind === 'amd')
            return this._gpuAmd();
        if (this._gpuKind === 'intel')
            return this._gpuIntel();
        return [null, null, null];
    }

    _detectGpu() {
        if (GLib.find_program_in_path('nvidia-smi'))
            return 'nvidia';
        return this._detectGpuSysfs();
    }

    _detectGpuSysfs() {
        for (const card of listDir('/sys/class/drm')) {
            if (!/^card\d+$/.test(card))
                continue;
            if (exists(`/sys/class/drm/${card}/device/gpu_busy_percent`))
                return 'amd';
        }
        for (const card of listDir('/sys/class/drm')) {
            if (!/^card\d+$/.test(card))
                continue;
            const base = `/sys/class/drm/${card}/device/hwmon`;
            for (const h of listDir(base)) {
                if (listDir(`${base}/${h}`).some(e => /^temp\d+_input$/.test(e)))
                    return 'intel';
            }
        }
        return 'none';
    }

    async _gpuNvidia() {
        try {
            const proc = Gio.Subprocess.new(
                ['nvidia-smi',
                    '--query-gpu=utilization.gpu,temperature.gpu,name',
                    '--format=csv,noheader,nounits'],
                Gio.SubprocessFlags.STDOUT_PIPE | Gio.SubprocessFlags.STDERR_SILENCE);
            const [out] = await proc.communicate_utf8_async(null, null);
            const line = (out ?? '').trim().split('\n')[0];
            if (!line)
                return [null, null, null];
            const [loadS, tempS, ...nameParts] = line.split(',');
            const load = parseFloat(loadS);
            const temp = parseFloat(tempS);
            return [
                Number.isNaN(load) ? null : load,
                Number.isNaN(temp) ? null : temp,
                nameParts.join(',').trim() || null,
            ];
        } catch {
            return [null, null, null];
        }
    }

    _gpuAmd() {
        let load = null, temp = null;
        for (const card of listDir('/sys/class/drm')) {
            if (!/^card\d+$/.test(card))
                continue;
            const dev = `/sys/class/drm/${card}/device`;
            const v = readInt(`${dev}/gpu_busy_percent`);
            if (v === null)
                continue;
            load = v;
            for (const h of listDir(`${dev}/hwmon`)) {
                for (const entry of listDir(`${dev}/hwmon/${h}`)) {
                    if (!/^temp\d+_input$/.test(entry))
                        continue;
                    const raw = readInt(`${dev}/hwmon/${h}/${entry}`);
                    if (raw !== null && raw / 1000 > 0 && raw / 1000 < 150) {
                        temp = raw / 1000;
                        break;
                    }
                }
                if (temp !== null)
                    break;
            }
            break;
        }
        return [load, temp, 'AMD GPU'];
    }

    _gpuIntel() {
        for (const card of listDir('/sys/class/drm')) {
            if (!/^card\d+$/.test(card))
                continue;
            const base = `/sys/class/drm/${card}/device/hwmon`;
            for (const h of listDir(base)) {
                for (const entry of listDir(`${base}/${h}`)) {
                    if (!/^temp\d+_input$/.test(entry))
                        continue;
                    const raw = readInt(`${base}/${h}/${entry}`);
                    if (raw !== null && raw / 1000 > 0 && raw / 1000 < 150)
                        return [null, raw / 1000, 'Intel GPU'];
                }
            }
        }
        return [null, null, 'Intel GPU'];
    }
}
